Use dataclass for SsdpServiceInfo

epenet opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Ssdp discovery now uses a dataclass instead of a dictionary object.
Access using `__getitem__` will fail in version 2022.6.
Please use `<cls>.<name>` or `<cls>.upnp.<name>` instead.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Ssdp discovery now uses a dataclass instead of a dictionary object.
Access using `__getitem__` will fail in version 2022.6.
Please use `<cls>.<name>` or `<cls>.upnp.<name>` instead.

Linked to this comment in the architecture discussion https://github.com/home-assistant/architecture/discussions/662#discussioncomment-1674270

## This is a central PR to see what needs to be done:
1. Revert incorrect `SsdpServiceInfo(TypedDict)`: 
- #60270
2. Rename matching domains as the current value prevents use as property :
-  #60283
3. Introduce new `SsdpServiceInfo(BaseServiceInfo)` class: 
- #60284
4. Update all tests in components to use the new `SsdpServiceInfo(BaseServiceInfo)` class: 
- basic: #60320
- `ssdp_location` needs to be optional: #60322
- `get` needs to be implemented: #60334
- `__iter__` needs to be implemented: #60339
- fix an issue with `in` checks #60340
5. (this PR) Update ssdp to use the new `SsdpServiceInfo(BaseServiceInfo)` in `discovery_flow.async_create_flow`
6. Update all components to use the new `SsdpServiceInfo` in `async_step_ssdp`

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

epenet committed:
Create SsdpServiceInfo

epenet committed:
Adjust tests

epenet commented:
Should `info[ATTR_UPNP_UDN]` also be duplicated/moved to `info[ATTR_UPNP][ATTR_UPNP_UDN]`?

bdraco commented:
> Should `info[ATTR_UPNP_UDN]` also be duplicated/moved to `info[ATTR_UPNP][ATTR_UPNP_UDN]`?

Only if it can have arbitrary keys. I think the answer is yes, but it would be good to get a comment from StevenLooman or chishm to confirm

epenet commented:
> 
> > Should `info[ATTR_UPNP_UDN]` also be duplicated/moved to `info[ATTR_UPNP][ATTR_UPNP_UDN]`?
> 
> Only if it can have arbitrary keys. I think the answer is yes, but it would be good to get a comment from StevenLooman or chishm to confirm

cc @StevenLooman

epenet committed:
Also copy ATTR_UPNP_UDN

epenet commented:
In order to help with the review (I hope) I have added to this PR the corresponding changes to `dlna_dmr` component as an example.

If you feel that it makes the PR too big and it is better to move it back out into a separate PR I will be happy to oblige.

chishm commented:
I'll take a closer look when I've got more time, but the biggest concern I have at the moment is that, for discovery,  components often match UPnP info (e.g. `deviceType`) in their manifest.json. That will possibly break if it's not in the top-level discovery info dict.

epenet commented:
> I'll take a closer look when I've got more time, but the biggest concern I have at the moment is that, for discovery, components often match UPnP info (e.g. `deviceType`) in their manifest.json. That will possibly break if it's not in the top-level discovery info dict.

@chishm this PR is not really a breaking change, as the data is duplicated for now. It is more a warning to other users to show that the real breaking change is coming in a future release.
I will prepare anyway the removal PR so that we can track and test it.

epenet commented:
I have created the draft removal PR: #59960

chishm commented:
> @chishm this PR is not really a breaking change, as the data is duplicated for now. It is more a warning to other users to show that the real breaking change is coming in a future release. I will prepare anyway the removal PR so that we can track and test it.

Understood, but I thought it best to flag early that this will break how discovery info is matched to components. For example, arcam_fmj defines in it's manifest.json the following:
```json
  "ssdp": [
    {
      "deviceType": "urn:schemas-upnp-org:device:MediaRenderer:1",
      "manufacturer": "ARCAM"
    }
  ]
```

In the ssdp component this is matched by `IntegrationMatchers.async_matching_domains` to the UPnP info in the `combined_headers` dict. The match is only done at the top level of the dict (no recursion), and is optimised for quickly checking SSDP notifications/searches against all domains.

This will need more consideration before altering the structure of the `combined_headers` dict.

epenet commented:
> Understood, but I thought it best to flag early that this will break how discovery info is matched to components. 

Yes it is already covered in the tests, and implemented in draft https://github.com/home-assistant/core/pull/59960

chishm commented:
> Yes it is already covered in the tests, and implemented in draft #59960

Ah, thank you for pointing that out, I missed it on first read-through. I withdraw my concern and will give this a full review tomorrow :smile:

epenet committed:
Use dataclass for SsdpServiceInfo

epenet committed:
Fix mypy

epenet commented on homeassistant/components/dlna_dmr/config_flow.py:
```diff
@@ -205,7 +204,7 @@ async def async_step_import_turn_on(
         return self.async_show_form(step_id="import_turn_on", errors=errors)
 
-    async def async_step_ssdp(self, discovery_info: DiscoveryInfoType) -> FlowResult:
+    async def async_step_ssdp(self, discovery_info: ssdp.SsdpServiceInfo) -> FlowResult:  # type: ignore[override]
```
`# type: ignore[override]` is temporary (required by mypy), and will be removed in part 2

chishm commented on homeassistant/components/ssdp/__init__.py:
```diff
@@ -85,56 +87,60 @@
 
     ssdp_usn: str
     ssdp_st: str
+    ssdp_location: str | None = None
```
This being optional is good. The location won't always be available in ssdp:byebye messages, which are potentially handled by functions of type `SsdpCallback` registered via `ssdp.async_register_callback`.

epenet commented:
Temporary rebase on dev - still waiting on #60339

bdraco commented:
Needs another rebase

epenet committed:
Rebased on dev

epenet committed:
Fix dlna_dmr

epenet committed:
Adjust async_step_ssdp signature

epenet committed:
Reduce size of PR

epenet committed:
ssdp discovery data cannot be None

epenet committed:
Fiw test_config_entries

epenet committed:
Adjust tests

epenet committed:
Fix typo

epenet commented on homeassistant/components/ssdp/__init__.py:
```diff
@@ -100,19 +88,20 @@ class _SsdpServiceDescription:
 
     ssdp_usn: str
-    ssdp_st: str
+    ssdp_st: str | None = None
```
I have made this optional because in my opinion this is not full proof (maybe both `st` and `nt` are None):
```python
    # Increase compatibility.
    if not info.ssdp_st and info.ssdp_nt:
        info.ssdp_st = info.ssdp_nt
```

chishm replied:
Depending on the message type, either the ST (Search Target, from M-SEARCH messages) or NT (Notification Type, from NOTIFY messages) header is mandatory. So you'll always have one or the other, but probably not both. Their contents are pretty much identical too.

epenet replied:
I have adjusted the code accordingly.

epenet commented on homeassistant/components/ssdp/__init__.py:
```diff
@@ -565,21 +570,30 @@ async def async_get_discovery_info_by_udn(self, udn: str) -> list[dict[str, Any]
+        ssdp_usn=combined_headers["usn"],
+        ssdp_ext=combined_headers.get("ext"),
+        ssdp_server=combined_headers.get("server"),
+        ssdp_st=combined_headers.get("st"),
```
If we decide that ssdp_st cannot be None, then I can adjust here to this, but it could generate a key error:
```suggestion
        ssdp_st=combined_headers.get("st", combined_headers["nt"]),
```

bdraco replied:
@chishm has more experience with st vs nt messages.  Probably best to wait for his thoughts

epenet committed:
Set ssdp_st as required

bdraco commented on homeassistant/components/ssdp/__init__.py:
```diff
@@ -565,23 +570,36 @@ async def async_get_discovery_info_by_udn(self, udn: str) -> list[dict[str, Any]
+) -> SsdpServiceInfo:
     """Convert headers and description to discovery_info."""
-    info = {
-        DISCOVERY_MAPPING.get(k.lower(), k): v
```
AFAICT `DISCOVERY_MAPPING` is now unused and should be dropped

epenet committed:
Drop DISCOVERY_MAPPING constant

epenet committed:
Replace __iter__ with __contains__ method

epenet commented:
I have dropped the `DISCOVERY_MAPPING` constant.

I have also dropped the `__iter__` method to replace it with `__contains__` method as I think it is a better way to solve the `in`/`not in` checks that some components still use.

bdraco commented:
I'll take a look tomorrow after some sleep.

bdraco commented:
I've run though this a few times now, and I think its good to go.  I'm going to test sonos with it first though

bdraco commented:
Sonos looks good.

Already tested discovery.
